src/components/workspace/editorMode/source.js

```diff
@@ -0,0 +1,831 @@
+                    next: "regex",
+                    regex: "]",
+                    token: "constant.language.escape",
+                }, {
+                    regex: "-",
+                    token: "constant.language.escape",
+                }, {
+                    next: "no_regex",
+                    regex: "$",
+                    token: "empty",
+                }, {
+                    defaultToken: "string.regexp.charachterclass"
+                }
+            ],
+            "start": [
+                DocCommentHighlightRules.getStartRule("doc-start"),
+                comments("start"),
+                {
+                    next: "regex",
+                    regex: "\\/",
+                    token: "string.regexp",
+        
+                }, {
+                    next : "start",
+                    regex : "\\s+|^$",
+                    token : "text",
+                    
+                }, {
+                    next: "no_regex",
+                    regex: "",
+                    token: "empty",
+                }
+            ],
+        };
+    
+    
+        if (!options || !options.noES6) {
+            this.$rules.no_regex.unshift({
+                nextState: "start",
+                onMatch: function(val, state, stack) {
+                    this.next = val === "{" ? this.nextState : "";
+                    if (val === "{" && stack.length) {
+                        stack.unshift("start", state);
+                    }
+                    else if (val === "}" && stack.length) {
+                        stack.shift();
+                        this.next = stack.shift();
+                        if (this.next.indexOf("string") !== -1 || this.next.indexOf("jsx") !== -1) {
+                            return "paren.quasi.end";
+                        }
+                    }
+                    return val === "{" ? "paren.lparen" : "paren.rparen";
+                },
+                regex: "[{}]", 
+            }, {
+                push  : [{
+                    regex : escapedRe,
+                    token : "constant.language.escape",
+                }, {
+                    push  : "start",
+                    regex : /\${/,
+                    token : "paren.quasi.start",
+                }, {
+                    next  : "pop",
+                    regex : /`/,
+                    token : "string.quasi.end",
+                }, {
+                    defaultToken: "string.quasi"
+                }],
+                regex : /`/,
+                token : "string.quasi.start",
+            });
+    
+            if (!options || options.jsx !== false) {
+                JSX.call(this);
+            }
+        }
+    
+        this.embedRules(DocCommentHighlightRules, "doc-",
+            [ DocCommentHighlightRules.getEndRule("no_regex") ]);
+    
+        this.normalizeRules();
+    };
+    
+    oop.inherits(JavaScriptHighlightRules, TextHighlightRules);
+    
+    function JSX() {
```

Comment:
This method is very long (100 lines!). I understand that it seems to decide some rules for JSX syntax, but could this be stated explicitly?

Reply:
I copied this directly from the brace mode library. I am still trying to figure out how it works. Can we merge it first and save optimisation for later?

Reply:
Could you include a link to the original source? It might be helpful if someone were to look at this in the future